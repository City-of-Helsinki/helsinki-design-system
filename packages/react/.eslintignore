**/*.d.ts
webpack.config.js
.templates
lib/
**/*.stories.tsx
