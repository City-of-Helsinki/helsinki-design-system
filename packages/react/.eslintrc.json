{
  "parser": "@typescript-eslint/parser",
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:react/recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:import/typescript",
    "plugin:jest/recommended",
    "airbnb",
    "prettier/@typescript-eslint",
    "plugin:prettier/recommended" // this should always be the last item on this array
  ],
  "plugins": ["react", "@typescript-eslint", "prettier", "jest", "import"],
  "parserOptions": {
    "ecmaVersion": 2018,
    "ecmaFeatures": {
      "jsx": true
    }
  },
  "ignorePatterns": ["*.module.css"],
  "rules": {
    "no-use-before-define": "off",
    "@typescript-eslint/no-use-before-define": ["error"],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": ["**/*.test.ts", "**/*.test.tsx", "**/test*.ts*", "**/*.stories.tsx"]
      }
    ],
    "import/order": [
      "error",
      {
        "groups": ["builtin", "external", "internal", ["parent", "sibling", "index"]],
        "newlines-between": "always"
      }
    ],
    "react/jsx-filename-extension": "off",
    "react/jsx-one-expression-per-line": "off",
    "react/jsx-props-no-spreading": "off",
    "react/jsx-curly-newline": "off",
    "@typescript-eslint/explicit-function-return-type": "off",
    "react/prop-types": "off",
    "import/extensions": "off",
    "import/prefer-default-export": "off",
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "components": ["Label"],
        "required": {
          "some": ["nesting", "id"]
        },
        "allowChildren": false
      }
    ],
    "@typescript-eslint/explicit-member-accessibility": "off",
    "react/state-in-constructor": [2, "never"],
    "no-unused-expressions": [2, {
      "allowTernary": true
    }],
    "no-unused-vars": "off",
    "@typescript-eslint/no-unused-vars": ["error"]
  },
  "settings": {
    "react": {
      "version": "detect"
    },
    "import/resolver": {
      "node": {
        "extensions": [".js", ".jsx", ".ts", ".tsx"],
        "paths": ["src"]
      }
    }
  },
  "env": {
    "browser": true,
    "jest": true
  },
  "globals": {
    "JSX": true
  }
}
