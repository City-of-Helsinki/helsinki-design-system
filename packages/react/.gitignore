node_modules/
storybook-static/
lib/
_bundles/
*.css.d.ts
.loki/current
.loki/difference

*storybook.log
