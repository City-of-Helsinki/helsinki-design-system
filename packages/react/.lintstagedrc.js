module.exports = {
  'src/**/*.{ts,tsx}': 'eslint',
}
