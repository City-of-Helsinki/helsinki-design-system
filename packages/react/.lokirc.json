{
  "chromeRetries": 2,
  "configurations": {
    "chrome.laptop": {
      "target": "chrome.docker",
      "width": 1366,
      "height": 768,
      "mobile": false
    },
    "chrome.iphone7": {
      "target": "chrome.docker",
      "preset": "iPhone 7"
    }
  },
  "skipStories": "Icons/.*|Button/*|Tag/*|Hero/*|Koros/*|Checkbox/*",
  "pixelmatch": {
    "threshold": 0.2,
    "alpha": 1,
    "diffColor": [255, 0, 255]
  }
}
