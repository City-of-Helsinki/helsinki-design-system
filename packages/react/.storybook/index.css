@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/565d73a693abe0776c801607ac28f0bf.woff') format('woff');
  font-display: swap;
  font-weight: 400;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/5bb29e3b7b1d3ef30121229bbe67c3e1.woff') format('woff');
  font-display: swap;
  font-weight: 400;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/7c46f288e8133b87e6b12b45dac71865.woff') format('woff');
  font-display: swap;
  font-weight: 500;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/e62dc97e83a385e4d8cdc939cf1e4213.woff') format('woff');
  font-display: swap;
  font-weight: 500;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/533af26cf28d7660f24c2884d3c27eac.woff') format('woff');
  font-display: swap;
  font-weight: 700;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/20d494430c87e15e194932b729d48270.woff') format('woff');
  font-display: swap;
  font-weight: 700;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/a50a1bd245ce63abcc0d1da80ff790d2.woff') format('woff');
  font-display: swap;
  font-weight: 900;
  font-style: normal;
  text-rendering: optimizeLegibility;
}

@font-face {
  font-family: 'HelsinkiGrotesk';
  src: url('../src/fonts/62a1781d8b396fbb025b0552cf6304d2.woff') format('woff');
  font-display: swap;
  font-weight: 900;
  font-style: italic;
  text-rendering: optimizeLegibility;
}

body {
  font-family: var(--font-default);
  font-size: var(--fontsize-body-m);
  color: var(--color-black-90);
  line-height: var(--lineheight-l);
}

.password-input--external-show-password-button,
.date-input--external-clear-value-button {
  display: grid;
  grid-template-columns: 400px 200px;
  gap: 20px;
  align-items: center;
}

@media only screen and (max-width: 768px) {
  .password-input--external-show-password-button,
  .date-input--external-clear-value-button {
    grid-template-columns: 300px;
  }
}

@media only screen and (max-width: 575.98px) {
  .table-custom-action {
    width: 100%;
  }
}

/* STEPPER */

.stepper-small {
  max-width: 300px;
}

@media only screen and (max-width: 575.98px) {
  .stepper-small {
    max-width: 100%;
  }
}

.stepper-margin {
  margin-left: -35px;
}

.stepper-heading {
  margin-left: 10px;
}

.stepper-form-validation {
  margin-left: -18px;
}

.stepper-card > div:first-child {
  margin-bottom: var(--spacing-m) !important;
}

#focused-element-after-cookie-consent-closed:focus {
  outline: 2px black solid;
}
