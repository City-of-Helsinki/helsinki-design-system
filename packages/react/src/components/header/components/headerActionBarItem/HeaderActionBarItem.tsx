/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { useEffect, useRef, useState } from 'react';

import classes from './HeaderActionBarItem.module.scss';
import { HeaderActionBarItemButton, HeaderActionBarItemButtonProps } from './HeaderActionBarItemButton';
import { useHeaderContext } from '../../HeaderContext';
import classNames from '../../../../utils/classNames';
import { AllElementPropsWithRef } from '../../../../utils/elementTypings';

export type HeaderActionBarItemProps = React.PropsWithChildren<
  AllElementPropsWithRef<'div'> & {
    /**
     * An avatar which replaces the icon. Usually user's initials, but can be any Element.
     */
    avatar?: string | JSX.Element;
    /**
     * Icon for the action bar item when dropdown is open.
     */
    closeIcon?: JSX.Element;
    /**
     * Label for the action bar item when dropdown is open.
     */
    closeLabel?: string | JSX.Element;
    /**
     * Additional classname for the dropdown element.
     */
    dropdownClassName?: string;
    /**
     * Fix the item position to the right side of the action bar.
     */
    fixedRightPosition?: boolean;
    /**
     * Possibility to use a full-width version of the dropdown, for example in mobile use.
     */
    fullWidth?: boolean;
    /**
     * Icon for the action bar item.
     */
    icon?: JSX.Element;
    /**
     * Additional classname for the icon.
     */
    iconClassName?: string;
    /**
     * ID of the dropdown item.
     */
    id: string;
    /**
     * Label for the action bar item.
     */
    label?: string | JSX.Element;
    /**
     * Positions the label right side of the icon.
     */
    labelOnRight?: boolean;
    /**
     * Menu button resizing is prevented by rendering button's active state to a separate element.
     */
    preventButtonResize?: boolean;
  }
>;

export const HeaderActionBarItem = (properties: HeaderActionBarItemProps) => {
  const {
    id,
    children,
    label,
    fullWidth,
    className: classNameProp,
    iconClassName: iconClassNameProp,
    dropdownClassName: dropdownClassNameProp,
    closeLabel,
    icon,
    closeIcon,
    'aria-label': ariaLabel,
    labelOnRight,
    fixedRightPosition,
    preventButtonResize,
    avatar,
    ...rest
  } = properties;
  const dropdownContentElementRef = useRef<HTMLDivElement>(null);
  const containerElementRef = useRef<HTMLDivElement>(null);
  const [hasContent, setHasContent] = useState(false);
  const { isSmallScreen } = useHeaderContext();
  const [visible, setDisplayProperty] = useState(false);

  const getContainer = () => containerElementRef.current as HTMLDivElement;

  const handleButtonClick = () => {
    setDisplayProperty(!visible);
  };

  const handleDocumentClick = (event) => {
    if (!visible) return;
    const container = getContainer();
    const eventTargetNode = event.target;
    if (!container.contains(eventTargetNode)) {
      setDisplayProperty(false);
    }
  };

  const handleBlur = (event) => {
    if (!visible) return;
    const container = getContainer();
    const eventTargetNode = event.relatedTarget;
    // close the dropdown if the focus is outside the container on large screens
    if (!container.contains(eventTargetNode) && !isSmallScreen) {
      setDisplayProperty(false);
    }
  };

  useEffect(() => {
    document.addEventListener('click', handleDocumentClick);
    return () => document.removeEventListener('click', handleDocumentClick);
  }, [containerElementRef.current]);

  // Hide the component if there is no content
  useEffect(() => {
    setHasContent(dropdownContentElementRef.current?.childNodes.length !== 0);
  }, [children]);

  const iconLabel = closeLabel && visible && !preventButtonResize ? closeLabel : label;
  const iconClass = closeIcon && visible && !preventButtonResize ? closeIcon : icon;
  const hasSubItems = React.Children.count(children) > 0;
  const visibilityClasses = {
    visible,
    [classes.visible]: visible,
    [classes.hasContent]: hasContent,
    [classes.fullWidth]: fullWidth || (isSmallScreen && fixedRightPosition),
    [classes.hasSubItems]: hasSubItems,
    [classes.menuItem]: id === 'Menu',
  };
  const className = classNames(classes.container, classNameProp, visibilityClasses);
  const iconClassName = classNames(classes.icon, iconClassNameProp);
  const dropdownClassName = classNames(classes.dropdown, dropdownClassNameProp, visibilityClasses);
  const buttonOverlayProps: Pick<HeaderActionBarItemButtonProps, 'activeStateIcon' | 'activeStateLabel'> =
    preventButtonResize
      ? {
          activeStateIcon: closeIcon,
          activeStateLabel: closeLabel,
        }
      : {};

  /* eslint-disable jsx-a11y/no-noninteractive-tabindex */
  return (
    <div {...rest} id={id} className={className} ref={containerElementRef} onBlur={handleBlur}>
      <HeaderActionBarItemButton
        className={iconClassName}
        onClick={handleButtonClick}
        label={iconLabel}
        icon={iconClass}
        aria-expanded={visible}
        aria-label={ariaLabel !== undefined ? ariaLabel : String(label)}
        aria-controls={id === 'Menu' ? `hds-mobile-menu` : `${id}-dropdown`}
        labelOnRight={labelOnRight}
        fixedRightPosition={fixedRightPosition}
        isActive={visible}
        fullWidth={fullWidth}
        hasSubItems={hasSubItems}
        avatar={avatar}
        preventButtonResize={preventButtonResize}
        {...buttonOverlayProps}
      />
      {hasSubItems && (
        <div className={classes.dropdownWrapper}>
          <div id={`${id}-dropdown`} className={dropdownClassName} ref={dropdownContentElementRef}>
            {visible && !fullWidth && label && <h3>{label}</h3>}
            <ul>{children}</ul>
          </div>
        </div>
      )}
    </div>
  );
};

HeaderActionBarItem.defaultProps = {
  fullWidth: false,
};
