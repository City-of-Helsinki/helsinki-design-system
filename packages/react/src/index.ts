export * from './icons';
export * from './components';
export * from './ssr';
