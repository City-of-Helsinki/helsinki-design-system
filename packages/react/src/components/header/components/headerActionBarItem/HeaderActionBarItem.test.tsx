import React, { useRef } from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { HeaderActionBarItem, HeaderActionBarItemProps } from './HeaderActionBarItem';

const clickHandler = jest.fn();

const defaultLabel = 'button label';
const fakeIconTestId = 'fake-icon-test-id';
const defaultIcon = <span data-testid={fakeIconTestId} />;
const activeStateLabel = 'close label';
const activeStateIconTestId = 'close-icon-test-id';
const activeStateIcon = <span data-testid={activeStateIconTestId} />;
const ariaControlsId = 'aria-controls-id';
const ariaLabel = 'aria label';

const defaultProps: HeaderActionBarItemProps = {
  onClick: clickHandler,
  label: defaultLabel,
  icon: defaultIcon,
  'aria-expanded': false,
  'aria-label': ariaLabel,
  'aria-controls': ariaControlsId,
  labelOnRight: false,
  fixedRightPosition: false,
  isActive: false,
};

const activeStateProps: Pick<HeaderActionBarItemProps, 'activeStateIcon' | 'activeStateLabel'> = {
  activeStateLabel,
  activeStateIcon,
};

const RenderTestScenario = (props?: Partial<HeaderActionBarItemProps> & JSX.IntrinsicElements['button']) => {
  const ref = useRef<HTMLButtonElement>(null);
  const combinedProps = { ...defaultProps, ...props };
  return <HeaderActionBarItem {...combinedProps} ref={ref} data-testid="button" />;
};

describe('HeaderActionBarItem spec', () => {
  it('renders the component correctly and adds all classNames from props', () => {
    const { asFragment } = render(
      <RenderTestScenario labelOnRight fixedRightPosition className="extra-class" {...activeStateProps} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not have basic accessibility issues', async () => {
    const { container } = render(<RenderTestScenario />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('Label and icon are set', async () => {
    const { getByText, getAllByTestId } = render(<RenderTestScenario />);
    expect(() => getByText(defaultLabel)).not.toThrow();
    expect(() => getAllByTestId(fakeIconTestId)).not.toThrow();
  });
  it('Aria attributes are set', async () => {
    const { getByTestId } = render(<RenderTestScenario aria-expanded />);
    const button = getByTestId('button') as HTMLElement;
    expect(button.getAttribute('aria-label')).toBe(ariaLabel);
    expect(button.getAttribute('aria-controls')).toBe(ariaControlsId);
    expect(button.getAttribute('aria-expanded')).toBe('true');
  });
  it('ActiveState content is rendered when "isActive" is true and "activeStateLabel" or "activeStateIcon" are passed', async () => {
    const { getAllByTestId } = render(<RenderTestScenario isActive {...activeStateProps} />);
    expect(() => getAllByTestId(activeStateIconTestId)).not.toThrow();
  });
  it('ActiveState content is also rendered when isActive is false and "activeStateLabel" or "activeStateIcon" are passed', async () => {
    const { getByText, getAllByTestId } = render(<RenderTestScenario {...activeStateProps} />);
    expect(() => getByText(defaultLabel)).not.toThrow();
    expect(() => getAllByTestId(fakeIconTestId)).not.toThrow();
  });
  it('ActiveState content is not rendered even if isActive is true, but neither "activeStateLabel" or "activeStateIcon" are passed', async () => {
    const { getByText, getAllByTestId } = render(<RenderTestScenario isActive />);
    expect(() => getByText(defaultLabel)).not.toThrow();
    expect(() => getAllByTestId(fakeIconTestId)).not.toThrow();
    expect(() => getByText(activeStateLabel)).toThrow();
    expect(() => getAllByTestId(activeStateIconTestId)).toThrow();
  });
});
