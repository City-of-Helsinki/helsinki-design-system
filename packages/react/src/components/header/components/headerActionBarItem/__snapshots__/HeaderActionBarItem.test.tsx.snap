// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderActionBarItem spec renders the component correctly and adds all classNames from props 1`] = `
<DocumentFragment>
  <button
    aria-controls="aria-controls-id"
    aria-expanded="false"
    aria-label="aria label"
    class="actionBarItem extra-class fixedRightPosition"
    data-testid="button"
    type="button"
  >
    <span>
      <span
        class="actionBarItemIcon labelOnRight"
      >
        <span
          aria-hidden="true"
          data-testid="fake-icon-test-id"
        />
      </span>
      <span
        class="actionBarItemLabel labelOnRight"
      >
        button label
      </span>
    </span>
  </button>
</DocumentFragment>
`;
